components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadMiddleware.kt

```diff
@@ -7,34 +7,100 @@ package mozilla.components.feature.downloads
 ) : Middleware<BrowserState, BrowserAction> {
+    private val logger = Logger("DownloadMiddleware")
 
+    private var scope = CoroutineScope(coroutineContext)
+
+    @InternalCoroutinesApi
     override fun invoke(
         store: MiddlewareStore<BrowserState, BrowserAction>,
         next: (BrowserAction) -> Unit,
         action: BrowserAction
     ) {
-        next(action)
         when (action) {
             is DownloadAction.AddDownloadAction -> {
-                val intent = Intent(applicationContext, downloadServiceClass)
-                intent.putExtra(DownloadManager.EXTRA_DOWNLOAD_ID, action.download.id)
-                applicationContext.startService(intent)
+                next(action)
+                scope.launch {
+                    if (!action.restored) {
+                        downloadStorage.add(action.download)
+                        logger.debug("Added download ${action.download.fileName} to the storage")
+                    }
+                }
+                if (action.download.status !in arrayOf(COMPLETED, CANCELLED)) {
+                    val intent = Intent(applicationContext, downloadServiceClass)
+                    intent.putExtra(DownloadManager.EXTRA_DOWNLOAD_ID, action.download.id)
+                    applicationContext.startService(intent)
+                    logger.debug("Sending download intent ${action.download.fileName}")
+                }
+            }
+
+            is DownloadAction.RemoveDownloadAction -> {
+                scope.launch {
+                    store.state.downloads[action.downloadId]?.let {
+                        downloadStorage.remove(it)
+                        logger.debug("Removed download ${it.fileName} from the storage")
+                    }
+                }
+            }
+
+            is DownloadAction.UpdateDownloadAction -> {
+                val updated = action.download
+                store.state.downloads[updated.id]?.let { old ->
+                    // To not overwhelm the storage, we only send updates that are relevant,
+                    // we only care about properties, that we are stored on the storage.
+                    if (!DownloadStorage.areTheSame(old, updated)) {
+                        scope.launch {
+                            downloadStorage.update(action.download)
+                        }
+                        logger.debug("Updated download ${action.download.fileName} on the storage")
+                    }
+                }
+            }
+
+            is DownloadAction.RestoreDownloadsState -> {
+                scope.launch {
+                    downloadStorage.getDownloads().collect { downloads ->
+                        downloads.forEach { download ->
+                            if (!store.state.downloads.containsKey(download.id)) {
+                                store.dispatch(DownloadAction.AddDownloadAction(download, true))
+                                logger.error("Download restarted from db")
```

Comment:
This shouldn't be logged as `error` and should probably be restored, not restarted?

Reply:
Agree, It's a leftover from testing.